### PR TITLE
Fixes for glob indexing.

### PR DESCRIPTION
**Pull request by kshakir**
In the FileEvaluator, for arrays/maps, evaluate files from the left and right side.
For binary operators, don't evaluate the right if the left fails.

### CODE REVIEW

File: src/main/scala/wdl4s/expression/FileEvaluator.scala [hunk 1]
```diff
@@ -115,7 +115,7 @@ case class FileEvaluator(valueEvaluator: ValueEvaluator, coerceTo: WdlType = Wdl
       case a: Ast if a.isArrayOrMapLookup =>
         evalValue(a) match {
           case Success(f: WdlFile) => Success(Seq(f))
-          case _ => evaluateRecursive(a.getAttribute("rhs"))
+          case _ => evaluateRecursive(a.getAttribute("lhs"))
```
Comment (mcovarr) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: If there's an expression on the rhs that's no longer going to be evaluated?  Why don't we want to evaluate both sides?

Comment (mcovarr) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: e.g. `File globFile = glob("globFile.txt")[count - 1]`

Comment (kshakir) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: Based on the comment [above](https://github.com/broadinstitute/wdl4s/blob/e040cf832588aea71fd5e4eaa0e05a4b397c4bd2/src/main/scala/wdl4s/expression/FileEvaluator.scala#L72-L87), I believe this method only supposed to be used during early evaluation, _before_ a task actually runs.

So `glob("globFile.txt")[count - 1]` should return a `Seq(WdlGlobFile("globFile.txt"))` for this method / case. Right?

Comment (mcovarr) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: shrug?  I'd like to know Centaur's opinion on this. 🙂

Comment (kshakir) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: Centaur last night said it was fine. You'll have to take my offline word for it though, as- as far as I know- we don't have a way of travis testing wdl4s PRs.

Comment (geoffjentry) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: Not at the moment at least. I've been going with the hope, pray, be ready to revert if necessary method.

Comment (geoffjentry) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: Also @kshakir I'm buying what you're selling. But I reserve the right to say that I sided with @mcovarr all along if travis winds up irritated!

**Comment by geoffjentry** [comment]
@kshakir is there a ticket associated w/ this? I'm mainly just interested to know the background

**Comment by geoffjentry** [comment]
👍

File: src/main/scala/wdl4s/expression/FileEvaluator.scala (hunk 1 above)
Comment (Horneth) [inline src/main/scala/wdl4s/expression/FileEvaluator.scala]: The only think I can think of that not evaluating the `rhs` would do is fail to detect if someone uses files as keys for a map...
If we really want to catch this case we could still do something like
```scala
evaluateRecursive(a.getAttribute("lhs")) flatMap { lhs => evaluateRecursive(a.getAttribute("rhs")) map { rhs => lhs ++ rhs } }
```

**Comment by Horneth** [comment]
👍

**Comment by kshakir** [comment]
Updated to evaluate left and right sides of map/array lookups.

File: src/test/scala/wdl4s/expression/FileEvaluatorSpec.scala [hunk 2]
```diff
@@ -45,12 +42,13 @@ class FileEvaluatorSpec extends FlatSpec with Matchers {
     (""" [read_string("x"), read_string("y")] """, Seq(WdlSingleFile("x"), WdlSingleFile("y")), WdlArrayType(WdlStringType)),
     (""" {read_int("a"): read_string("x"), 4: read_string("y")} """, Seq(WdlSingleFile("a"), WdlSingleFile("x"), WdlSingleFile("y")), WdlArrayType(WdlStringType)),
     (""" glob("out-*.txt") """, Seq(WdlGlobFile("out-*.txt")), WdlFileType),
+    (""" glob("out-*.txt")[0] """, Seq(WdlGlobFile("out-*.txt")), WdlFileType),
```
Comment (mcovarr) [inline src/test/scala/wdl4s/expression/FileEvaluatorSpec.scala]: can you have an `rhs` array or map indexing test here?

**Comment by mcovarr** [comment]
:+1: anew

File: src/test/scala/wdl4s/expression/FileEvaluatorSpec.scala (hunk 2 above)
Comment (kshakir) [inline src/test/scala/wdl4s/expression/FileEvaluatorSpec.scala]: I don't know of any that would actually pass parsing or generate a proper AST. Filing a new ticket.

Comment (kshakir) [inline src/test/scala/wdl4s/expression/FileEvaluatorSpec.scala]: https://github.com/broadinstitute/wdl4s/issues/90

**Pull request merged**